src/radiation/mocmc.cpp

```diff
@@ -0,0 +1,947 @@
+            Vec cov_Fstar{v(iblock, idx_F(ispec, 0), k, j, i) / sdetgam,
+                          v(iblock, idx_F(ispec, 1), k, j, i) / sdetgam,
+                          v(iblock, idx_F(ispec, 2), k, j, i) / sdetgam};
+            Tens2 con_tilPi;
+            SPACELOOP2(ii, jj) {
+              con_tilPi(ii, jj) = v(iblock, iTilPi(ispec, ii, jj), k, j, i);
+            }
+            Real JBB = opac_d.EnergyDensityFromTemperature(v(iblock, pT, k, j, i),
+                                                           dev_species[ispec]);
+
+            ClosureMOCMC<Vec, Tens2> c(con_v, &g);
+
+            Real dE = 0;
+            Vec cov_dF;
+
+            // Calculate Inu0
+            for (int bin = 0; bin < nu_bins; bin++) {
+              v(iblock, Inu0(ispec, bin), k, j, i) = 0.;
+            }
+            for (int n = 0; n < nsamp; n++) {
+              const int nswarm = swarm_d.GetFullIndex(k, j, i, n);
+              const Real dOmega = (mu_hi(nswarm) - mu_lo(nswarm)) *
+                                  (phi_hi(nswarm) - phi_lo(nswarm)) / (4. * M_PI);
+              for (int bin = 0; bin < nu_bins; bin++) {
+                v(iblock, Inu0(ispec, bin), k, j, i) +=
+                    Inuinv(bin, ispec, nswarm) * pow(nusamp(bin), 3) * dOmega;
+              }
+            }
+
+            // Frequency-average opacities
+            Real kappaJ = 0.;
+            Real kappaH = 0.;
+
+            Real Itot = 0.;
+            for (int bin = 0; bin < nu_bins; bin++) {
+              kappaJ += opac_d.AngleAveragedAbsorptionCoefficient(
+                            v(iblock, pdens, k, j, i), v(iblock, pT, k, j, i),
+                            v(iblock, pye, k, j, i), dev_species[ispec], nusamp(bin)) *
+                        v(iblock, Inu0(ispec, bin), k, j, i) * nusamp(bin);
+              Itot += v(iblock, Inu0(ispec, bin), k, j, i) * nusamp(bin);
+            }
+
+            // Trapezoidal rule
+            kappaJ -= 0.5 *
+                      opac_d.AngleAveragedAbsorptionCoefficient(
+                          v(iblock, pdens, k, j, i), v(iblock, pT, k, j, i),
+                          v(iblock, pye, k, j, i), dev_species[ispec], nusamp(0)) *
+                      v(iblock, Inu0(ispec, 0), k, j, i) * nusamp(0);
+            kappaJ -=
+                0.5 *
+                opac_d.AngleAveragedAbsorptionCoefficient(
+                    v(iblock, pdens, k, j, i), v(iblock, pT, k, j, i),
+                    v(iblock, pye, k, j, i), dev_species[ispec], nusamp(nu_bins - 1)) *
+                v(iblock, Inu0(ispec, nu_bins - 1), k, j, i) * nusamp(nu_bins - 1);
+            Itot -= 0.5 *
+                    (v(iblock, Inu0(ispec, 0), k, j, i) * nusamp(0) +
+                     v(iblock, Inu0(ispec, nu_bins - 1), k, j, i) * nusamp(nu_bins - 1));
+            kappaJ = robust::ratio(kappaJ, Itot);
+            kappaH = kappaJ;
+            kappaJ *= (1. - scattering_fraction);
```

Comment:
We should probably add at least a TODO here to remove the `scattering_fraction` correction. I could see it causing some issue in the future when we have physical scattering in.

Reply:
Agreed, `TODO` added. Adding a gray scattering opacity to `singularity-opac` is near the top of my todo list so hopefully I get to this soon.